assets/js/googlesitekit/modules/datastore/modules.js

```diff
@@ -154,14 +174,92 @@ const baseActions = {
 
 		return { response, error };
 	},
+
+	/**
+	 * Registers a module
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {string}          slug                         Module slug.
+	 * @param {Object}          settings                     Module settings.
+	 * @param {string}          [settings.name]              Optional. Module name. Default is the slug.
+	 * @param {string}          [settings.description]       Optional. Module description. Default empty string.
+	 * @param {string}          [settings.icon]              Optional. Module icon. Default empty string.
+	 * @param {number}          [settings.order]             Optional. Numeric indicator for module order. Default 10.
+	 * @param {string}          [settings.homepage]          Optional. Module homepage URL. Default empty string.
+	 * @param {WPElement}       [settings.settingsComponent] React component to render the settings panel. Default is the DefaultModuleSettings component.
+	 * @return {Object}                                      Redux-style action.
+	 */
+	*registerModule( slug, { settingsComponent = DefaultModuleSettings, ...settings } = {} ) {
+		invariant( slug, 'module slug is required' );
+
+		const registry = yield commonActions.getRegistry();
+		let registryKey = yield registry.select( CORE_SITE ).getRegistryKey();
+		if ( registryKey === undefined ) {
+			registryKey = uuidv4();
+			yield registry.dispatch( CORE_SITE ).setRegistryKey( registryKey );
+		}
+
+		// We do this assignment in the action rather than the reducer because we can't send a
+		// payload that includes a React component to the reducer; we'll get an error about
+		// payloads needing to be plain objects.
+		if ( ModuleComponents[ registryKey ] === undefined ) {
+			ModuleComponents[ registryKey ] = {};
+		}
+		if ( ModuleComponents[ registryKey ][ slug ] === undefined ) {
+			ModuleComponents[ registryKey ][ slug ] = settingsComponent;
+		}
+
+		const mergedModuleSettings = {
+			slug,
+			name: slug,
+			description: null,
+			icon: null,
+			order: 10,
+			homepage: null,
+			internal: false,
+			...settings,
+		};
+
+		return {
+			payload: { slug, settings: mergedModuleSettings },
+			type: REGISTER_MODULE,
+		};
+	},
 };
 
 export const baseControls = {
-	[ REFETCH_AUTHENICATION ]: () => {
+	[ REFETCH_AUTHENTICATION ]: () => {
 		return API.get( 'core', 'user', 'authentication', { timestamp: Date.now() }, { useCache: false } );
 	},
 };
 
+const baseReducer = ( state, { type, payload } ) => {
+	switch ( type ) {
+		case REGISTER_MODULE: {
+			const { modules: existingModules } = state;
+			const { slug, settings } = payload;
+			return {
+				...state,
+				modules: {
+					...existingModules,
+					[ slug ]: settings,
+				},
+			};
+		}
+		case SET_MODULE_COMPONENT_KEY: {
+			const { registryKey } = payload;
+			return {
+				...state,
+				registryKey,
+			};
+		}
```

Comment:
This doesn't seem to be used anymore now with the centralized registry approach. Remove?